xof/k12/k12.go

```diff
@@ -0,0 +1,372 @@
+// k12 implements the KangarooTwelve XOF.
+package k12
+
+import (
+	"encoding/binary"
+
+	"github.com/cloudflare/circl/internal/sha3"
+	"github.com/cloudflare/circl/simd/keccakf1600"
+)
+
+const chunkSize = 8192 // aka B
+
+// KangarooTwelve splits the message into chunks of 8192 bytes each.
+// The first chunk is absorbed directly in a TurboSHAKE128 instance, which
+// we call the stalk. The subsequent chunks aren't absorbed directly, but
+// instead their hash is absorbed: they're like leafs on a stalk.
+// If we have a fast TurboSHAKE128 available, we buffer chunks until we have
+// enough to do the parallel TurboSHAKE128. If not, we absorb directly into
+// a separate TurboSHAKE128 state.
+
+type State struct {
+	initialTodo int // Bytes left to absorb for the first chunk.
+
+	stalk sha3.State
```

Comment:
Adding SwitchDS will allow, for example, the SHA3 instance to modify the DS.
what about this:

```suggestion
	stalk sha3.TurboHash
```
where 
```
type TurboHash interface {
   ShakeHash
   SwitchDS(byte)
}
```

Reply:
`SwitchDS` is part of an internal package, so we're not exposing a dangerous API.
Interfaces are problematic because they typically force variables onto the heap.